### ISSUE TITLE
Provide support for bech32 encoding of address bytes (encode/decode)

### ISSUE THREAD

**Issue by ethanfrey** [iov-one/iov-core#339]
Bech32 is the future: https://en.bitcoin.it/wiki/Bech32
"Readable" addresses with error correction.

This would replace displaying hex-strings for addresses, not the human-names
Moved here from: https://github.com/iov-one/weave/issues/27

**Comment by ethanfrey** [comment]
I think this would be nice for displaying addresses with checksums for bns chain.
Just expose it in Encoding or Crypto libraries, so the clients can use it.

https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
https://github.com/bitcoinjs/bech32

We don't deal with in internally, but promote usage in the client.

**Comment by webmaster128** [comment]
Does it make sense to implement this before we support a coin from [SLIP-0173](https://github.com/satoshilabs/slips/blob/master/slip-0173.md)?

**Comment by Isabello** [comment]
I would be happy to use bech32 addresses. Theres is also this slip to go along with it: https://github.com/satoshilabs/slips/blob/master/slip-0173.md

We should probably decide on what identifier that we want on the front of the address

**Comment by Isabello** [comment]
Looks like you got to the slip just before me :)

> Does it make sense to implement this before we support a coin from [SLIP-0173](https://github.com/satoshilabs/slips/blob/master/slip-0173.md)?

Maybe we can provide support for these kinds of addresses in Weave and move away from hex type for the BNS.

**Comment by webmaster128** [comment]
I wonder who is responsible for the address format. If the wallet takes everything from a uniform interface, the codec in IOV-Core must be responsible to provide a printable address in string format. Otherwise the wallet either needs to do case distinction (which we want to avoid) or re-encodes Lisk addresses (e.g. `bech32(ascii("16313739661670634666L"))`

**Comment by ethanfrey** [comment]
Yes, maybe makes sense that iov-core exposes string address in external interface and hides that detail from wallet. Internally we use bytes, but the codec can handle the transform. Like key To Address.... We can have an encoder/decoder for bytes <-> string with common address formats

**Comment by ethanfrey** [comment]
It is actually super important that we can expose addresses in "native" format, so people can use their existing wallets and send tokens into the IOV wallet. Also, so we can send from an IOV wallet to another wallet that doesn't use BNS. (In this case, we enter the typical chain address instead of a value name)

You are right that the chain-specific encoding should be done in iov-core

**Comment by webmaster128** [comment]
Looking at #421, we can make this change UI only, i.e. on the current BOV networks without changing something in the backend. The only thing we need to adapt is to do a bech32->hex conversion in the BovFaucet.credit implementation, because the faucet expects hex.

**Comment by webmaster128** [comment]
@LucasIsasmendi first thing we need here is a is a wrapper class in @iov/encoding with

```ts
class Bench32 {
  public static encode(prefix: string, data: Uint8Array): string;
  public static decode(address: string): { readonly prefix: string, readonly data: Uint8Array };
}
```

this can be added and tested before we know how and when we use it

**Comment by webmaster128** [comment]
As discussed on Wednesday, we use the prefixes "iov" and "tiov" for mainnet and testnets. A leading "t" seems to be the winning convention at https://github.com/satoshilabs/slips/blob/master/slip-0173.md, so this is good. An IOV address can be used for BNS or other services on our blockchain
